stbtrace/nfs.st

```diff
@@ -0,0 +1,336 @@
+#
+# Copyright (c) 2019 by Delphix. All rights reserved.
+#
+
+from bcc import BPF
+from time import sleep
+from time import time
+import datetime
+import ctypes as ct
+import sys
+sys.path.append('/opt/delphix/server/bin')
+from bcchelper import BCCHelper
+
+# BPF txg program
+bpf_text = """
+#include <uapi/linux/ptrace.h>
+#include <linux/bpf_common.h>
+#include <uapi/linux/bpf.h>
+#include <linux/sunrpc/svc.h>
+#include "/opt/delphix/server/etc/bcc_helper.h"
+
+
+// nfsd4 definitions from fs/nfsd/xdr4.h
+#define u32 unsigned int
+#define u64 unsigned long long
+#define bool int
+
+typedef struct {
+        u32             cl_boot;
+        u32             cl_id;
+} clientid_t;
+
+typedef struct {
+        clientid_t      so_clid;
+        u32             so_id;
+} stateid_opaque_t;
+
+typedef struct {
+        u32                     si_generation;
+        stateid_opaque_t        si_opaque;
+} stateid_t;
+
+typedef struct {
+        stateid_t       rd_stateid;     /* request */
+        u64		rd_offset;      /* request */
+        u32             rd_length;      /* request */
+        int             rd_vlen;
+        struct file     *rd_filp;
+        bool            rd_tmp_file;
+
+        void            *rd_rqstp;      /* response */
+        void            *rd_fhp;        /* response */
+} nfsd4_read;
+
+#define NFS4_VERIFIER_SIZE      8
+typedef struct { char data[NFS4_VERIFIER_SIZE]; } nfs4_verifier;
+
+typedef struct {
+        stateid_t       wr_stateid;         /* request */
+        u64             wr_offset;          /* request */
+        u32             wr_stable_how;      /* request */
+        u32             wr_buflen;          /* request */
+        struct kvec     wr_head;
+        struct page **  wr_pagelist;        /* request */
+        u32             wr_bytes_written;   /* response */
+        u32             wr_how_written;     /* response */
+        nfs4_verifier   wr_verifier;        /* response */
+} nfsd4_write;
+
+// Definitions for this script
+#define READ_STR "read"
+#define WRITE_STR "write"
+#define NFSV3_STR "v3"
+#define NFSV4_STR "v4"
+#define OP_NAME_LEN 6
+#define VER_NAME_LEN 3
+// Max length for null terminated string with ipv4 literal address
+#define MAX_IP_STRING 16
+// Client ip is sometimes proceeded by "*," or ","
+#define CLIENT_PREFIX_LEN 2
+#define CLIENT_LEN (MAX_IP_STRING + CLIENT_PREFIX_LEN)
+#define SYNC_WRITE  1
+#define ASYNC_WRITE 0
+#define CACHED_READ 1
+#define NONCACHED_READ 0
+#define AXIS_NOT_APPLICABLE -1
+
+// Structure to hold thread local data
+typedef struct {
+    u64 ts;
```

Comment:
style nit: Looks like earlier in the script, you were indenting by 8 column tabs (or 8 spaces). Please make all indentation uniform.